File: app/src/main/java/tpcreative/co/qrscanner/common/services/RetrofitHelper.java
package tpcreative.co.qrscanner.common.services;
import java.io.IOException;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import tpcreative.co.qrscanner.common.api.RootAPI;

public class RetrofitHelper {
    /**
     * The CityService communicates with the json api of the city provider.
     */
    public RootAPI getCityService(String url) {
        final Retrofit retrofit = createRetrofit(url);
        return retrofit.create(RootAPI.class);
    }
    /**
     * This custom client will append the "username=demo" query after every request.
     */
    private OkHttpClient createOkHttpClient() {
        final OkHttpClient.Builder httpClient =
                new OkHttpClient.Builder();
        httpClient.addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                final Request original = chain.request();
                final HttpUrl originalHttpUrl = original.url();
                final HttpUrl url = originalHttpUrl.newBuilder()
                        .build();
                // Request customization: add request headers
                final Request.Builder requestBuilder = original.newBuilder()
                        .url(url);
                final Request request = requestBuilder.build();
                return chain.proceed(request);
            }
        });
        return httpClient.build();
    }
    /**
     * Creates a pre configured Retrofit instance
     */
    private Retrofit createRetrofit(String url) {
        return new Retrofit.Builder()
                .baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create()) // <- add this
                .client(createOkHttpClient())
                .build();
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/common/services/RetrofitHelper.kt
package tpcreative.co.qrscanner.common.services

import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import tpcreative.co.qrscanner.common.api.RootAPI

class RetrofitHelper {
    /**
     * The CityService communicates with the json api of the city provider.
     */
    fun getCityService(url: String?): RootAPI? {
        val retrofit = createRetrofit(url)
        return retrofit.create(RootAPI::class.java)
    }

    /**
     * This custom client will append the "username=demo" query after every request.
     */
    private fun createOkHttpClient(): OkHttpClient? {
        val httpClient = OkHttpClient.Builder()
        httpClient.addInterceptor { chain ->
            val original = chain.request()
            val originalHttpUrl = original.url()
            val url = originalHttpUrl.newBuilder()
                    .build()
            // Request customization: add request headers
            val requestBuilder = original.newBuilder()
                    .url(url)
            val request = requestBuilder.build()
            chain.proceed(request)
        }
        return httpClient.build()
    }

    /**
     * Creates a pre configured Retrofit instance
     */
    private fun createRetrofit(url: String?): Retrofit? {
        return Retrofit.Builder()
                .baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create()) // <- add this
                .client(createOkHttpClient())
                .build()
    }
}